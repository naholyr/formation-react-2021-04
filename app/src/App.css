body {
  margin: 0;
  padding: 0;
  font-family: "Gill Sans", "Gill Sans MT", Calibri, "Trebuchet MS", sans-serif;
  font-size: 16px;
}

h1 {
  font-size: 3rem;
  margin: 0;
  padding: 0;
  text-align: center;
  color: darkblue;
}
