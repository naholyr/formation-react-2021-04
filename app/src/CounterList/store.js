import { combineReducers, createStore } from "redux";

export const initialState = {
  counters: {
    nbCounters: 0,
  },
};

export const countersReducer = (state = initialState.counters, action) => {
  switch (action.type) {
    case "CHANGE_NB_COUNTERS":
      return {
        ...state,
        nbCounters: action.payload.nb,
      };
    default:
      return state;
  }
};

export const reducer = combineReducers({
  counters: countersReducer,
});

export const initStore = () => {
  return createStore(reducer);
};
