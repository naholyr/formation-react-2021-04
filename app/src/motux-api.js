/* globals io:readonly */ // loaded from Motux server

import {
  addTrial,
  setGame,
  setInputDisabled,
  setScores,
  startNewGame,
} from "./actions";

const HOST = process.env.REACT_APP_MOTUX_HOST;

const socket = io("ws://" + HOST);

// username => token
export const login = async (username, signal = undefined) => {
  const response = await fetch(`http://${HOST}/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username }),
    signal,
  });
  const { token } = await response.json();
  console.log(`Logged in as ${username}`);
  localStorage.setItem("token", token);
  return token;
};

// token => username
export const relogin = async (token, signal = undefined) => {
  const response = await fetch(
    `http://${HOST}/whoami?token=${encodeURIComponent(token)}`,
    {
      method: "GET",
      signal,
    }
  );
  const { username } = await response.json();
  console.log(`Re-logged in as ${username}`);
  return username;
};

export const loadGame = (token, dispatch) => {
  socket.emit("login", token, (game) => {
    console.log("Game ready", game);
    dispatch(setGame(game));
  });
};

export const logout = () => {
  socket.emit("logout");
};

export const tryWord = (string) => {
  socket.emit("tryWord", string);
};

export const listenToGameEvents = (dispatch) => {
  socket.on("disableInput", () => {
    dispatch(setInputDisabled(true));
  });

  socket.on("enableInput", () => {
    dispatch(setInputDisabled(false));
  });

  socket.on("failure", (code) => {
    // TODO display error
    console.error("Failure", code);
  });

  socket.on("winner", (username) => {
    // TODO display winner
    console.warn("Winner!", username);
  });

  socket.on("addTrial", (trial) => {
    dispatch(addTrial(trial));
  });

  socket.on("updateScores", (scores) => {
    dispatch(setScores(scores));
  });

  socket.on("wordLength", (length) => {
    dispatch(startNewGame(length));
  });
};
