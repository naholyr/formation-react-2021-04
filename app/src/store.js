import { combineReducers, createStore } from "redux";

export const initialState = {
  user: {
    name: null,
  },
  game: {
    trials: [
      {
        name: "Coco",
        word: [
          ["B", 0],
          ["O", 0],
          ["U", 1],
          ["L", 0],
          ["I", 1],
          ["M", 1],
          ["I", 1],
          ["Q", 0],
          ["U", 1],
          ["E", 1],
        ],
        win: false,
      },
      {
        name: "Nicolas",
        word: [
          ["B", 0],
          ["I", 1],
          ["J", 0],
          ["O", 0],
          ["U", 1],
          ["T", 1],
          ["I", 1],
          ["E", 2],
          ["R", 1],
          ["S", 1],
        ],
        win: false,
      },
      {
        name: "Coco",
        word: [
          ["T", 1],
          ["E", 1],
          ["L", 0],
          ["E", 1],
          ["V", 0],
          ["I", 1],
          ["S", 1],
          ["I", 1],
          ["O", 0],
          ["N", 1],
        ],
        win: false,
      },
      {
        name: "Nicolas",
        word: [
          ["A", 0],
          ["B", 0],
          ["R", 1],
          ["I", 1],
          ["C", 0],
          ["O", 0],
          ["T", 1],
          ["I", 1],
          ["E", 1],
          ["R", 1],
        ],
        win: false,
      },
    ],
    scores: [
      {
        name: "Renaud",
        score: 9,
      },
      {
        name: "test",
        score: 7,
      },
      {
        name: "Coco",
        score: 0,
      },
      {
        name: "Romain F",
        score: 0,
      },
      {
        name: "ben",
        score: 0,
      },
      {
        name: "Nicolas",
        score: 0,
      },
    ],
    wordLength: 10,
    inputDisabled: false,
  },
};

export const userReducer = (state = initialState.user, action) => {
  switch (action.type) {
    case "SET_USER":
      return { ...state, name: action.payload.name };
    default:
      return state;
  }
};

export const gameReducer = (state = initialState.game, action) => {
  switch (action.type) {
    default:
      return state;
  }
};

export const reducer = combineReducers({
  user: userReducer,
  game: gameReducer,
});

export const initStore = () => {
  return createStore(
    reducer,
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  );
};
