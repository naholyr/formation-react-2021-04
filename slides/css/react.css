@media print {
  .print-warning {
    display: none;
  }
}

.reveal h1,
.reveal h2,
.reveal h3,
.reveal h4,
.reveal h5,
.reveal h6 {
  text-transform: none;
}

.reveal code {
  background: rgba(200, 200, 200, 0.3);
  font-family: monospace;
}

.reveal pre {
  border: 1px solid #ccc;
}
.reveal pre code {
  background: white;
  color: black;
}

.reveal pre.stretch {
  width: 80% !important;
  left: 50% !important;
  margin-left: -40% !important;
  font-size: 0.9em;
}

.reveal .raw img,
.reveal img.raw {
  background: none !important;
  border: none !important;
  box-shadow: none !important;
}

.screen {
  display: initial;
}
.print {
  display: none;
}

.print-pdf .reveal .cleartext * {
  color: black;
}
.print-pdf .reveal .cleartext a {
  color: #333366;
}
.print-pdf .screen {
  display: none;
}
.print-pdf .print {
  display: initial;
}


/* fix margins of lists followed by title */
.reveal ul+h1, .reveal ul+h2, .reveal ul+h3, .reveal ul+h4, .reveal ul+h5, .reveal ul+h6,
.reveal ol+h1, .reveal ol+h2, .reveal ol+h3, .reveal ol+h4, .reveal ol+h5, .reveal ol+h6 {
  margin-top: 20px;
}



html.intro body:not(.print-pdf),
body.print-pdf section#intro {
  background-image: url(../images/logos/react.png);
  background-repeat: no-repeat;
  background-position: center 50%;
  background-color: #222222;
  background-size: contain;
}
html.intro body:not(.print-pdf) *,
body.print-pdf section#intro * {
  color: white;
}
html.intro body:not(.print-pdf) * a {
  color: #ccccff;
}
html.intro body:not(.print-pdf) section h1 {
  text-align: right;
}

html.end body:not(.print-pdf),
body.print-pdf section#goodbye {
  background-image: url(../images/logos/byte-club.png);
  background-repeat: no-repeat;
  background-position: center bottom;
  background-size: 80vh;
}
body.print-pdf section#goodbye {
  background-size: 300px;
}
body.print-pdf section img {
  max-height: 300px;
}
