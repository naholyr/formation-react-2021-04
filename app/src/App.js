// import { lazy } from "react";
import "./App.css";
import { LoginForm } from "./LoginForm/LoginForm";
import { Game } from "./Game/Game";
import { useSelector } from "react-redux";
// const Game = lazy(() =>
//   import("./Game/Game").then((m) => ({ default: m.Game }))
// );

function App() {
  // useSelector(selector : state => any)
  // - store.subscribe => à chaque update du store, on refait la sélection
  // - "faire la sélection" = selector(state actuel du store) => valeur retournée par le hook
  const authenticated = useSelector((state) => state.user.name !== null);

  const content = authenticated ? <Game /> : <LoginForm />;

  return (
    <>
      <h1>Motux</h1>
      {content}
    </>
  );
}

export default App;
