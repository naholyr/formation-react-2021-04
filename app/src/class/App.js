// import { lazy } from "react";
import "../App.css";
import { LoginForm } from "./LoginForm";
import { Game } from "../Game/Game";
import { connect } from "react-redux";
import { loadGame, login, logout, relogin } from "../motux-api";
import { setUser } from "../actions";
import { Component } from "react";
import { string } from "prop-types";
import { bool } from "prop-types";

class App extends Component {
  static propTypes = {
    username: string,
    loaded: bool.isRequired,
  };

  static defaultProps = {
    username: null,
  };

  render() {
    const { username, loaded } = this.props;

    const loading = !loaded && username !== null;
    const authenticated = loaded && username !== null;

    const content = loading ? (
      <div>Loadingâ€¦</div>
    ) : authenticated ? (
      <Game />
    ) : (
      <LoginForm />
    );

    return (
      <>
        <h1>Motux</h1>
        {content}
      </>
    );
  }

  componentDidMount() {
    this.handleLogin();
  }

  componentWillUnmount() {
    this.handleLogout();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.username !== this.props.username) {
      this.handleLogout();
      this.handleLogin();
    }
  }

  handleLogin() {
    const { username, dispatch } = this.props;
    const token = localStorage.getItem("token");

    // token && username => already fully identified
    // !token && !username => not yet filled login form
    if ((username && !token) || (token && !username)) {
      const controller = new AbortController();
      if (token) {
        // relogin (token => username)
        relogin(token, controller.signal).then((username) => {
          dispatch(setUser(username)); // state.user.name = null => username
        });
        loadGame(token, dispatch); // state.game.loaded = false => true
      } else {
        // login (username => token)
        login(username, controller.signal)
          .then((token) => {
            loadGame(token, dispatch);
          })
          .catch((err) => {
            console.error(err);
            logout();
          });
      }
      return () => {
        controller.abort();
        logout();
      };
    }
  }

  handleLogout() {
    // TODO
    logout();
  }
}

const mapStateToProps = (state) => ({
  username: state.user.name,
  loaded: state.game.loaded,
});

export default connect(mapStateToProps)(App);
