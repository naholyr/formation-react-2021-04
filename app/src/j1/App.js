// import logo from './logo.svg';
import { useEffect, useState } from "react";
import { resetNbCounters } from "./actions";
import "./App.css";
import { CounterList } from "./CounterList/CounterList";

const INITIAL_COUNT = 2;

function App() {
  const [counterListKey, setCounterListKey] = useState(1);
  // const [nbCount, setNbCount] = useState(INITIAL_COUNT);
  const [nbCount, setNb] = useState(
    window.store.getState().counters.nbCounters
  );
  useEffect(() => {
    const unsubscribe = window.store.subscribe(() => {
      setNb(window.store.getState().counters.nbCounters);
    });
    return () => {
      unsubscribe();
    };
  }, []);

  useEffect(() => {
    console.log("App", { nbCount, counterListKey });
  });

  const onClickReset = (e) => {
    console.log("App.onClickReset");
    e.preventDefault();
    setCounterListKey(counterListKey + 1);
    window.store.dispatch(resetNbCounters());
  };

  // const foo = (nbCount) => {
  //   console.log("App.foo", nbCount);
  //   setNbCount(nbCount);
  // };

  return (
    <div className="App">
      <h1>Compteurs {nbCount}</h1>
      <CounterList
        key={counterListKey}
        // onCountChange={foo}
        initialCount={INITIAL_COUNT}
      />
      <button onClick={onClickReset}>reset</button>
    </div>
  );
}

export default App;
