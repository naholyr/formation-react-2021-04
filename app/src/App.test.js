import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import App from "./App";
import { reducer, initialState } from "./store";
import { loadGame, login, relogin } from "./motux-api";
import userEvent from "@testing-library/user-event";

jest.mock("./motux-api", () => {
  return {
    loadGame: jest.fn(),
    login: jest.fn(),
    relogin: jest.fn(),
    logout: jest.fn(),
  };
});

const renderWithStore = async (element, state = initialState) => {
  const store = createStore(reducer, state);
  const result = await render(<Provider store={store}>{element}</Provider>);
  return { ...result, store };
};

describe("App", () => {
  test("shows login form", () => {
    renderWithStore(<App />);
    const input = screen.getByPlaceholderText("Your name");
    expect(input).toBeInTheDocument();
  });

  // Snapshot DOM result
  test("snapshot", async () => {
    const { container } = await renderWithStore(<App />);
    expect(container).toMatchSnapshot();
  });

  afterEach(() => {
    localStorage.clear();
  });

  beforeEach(() => {
    login.mockResolvedValue("FAKE_TOKEN");
    relogin.mockResolvedValue("TOTO");
  });

  // Check auth process
  test("renders with username", async () => {
    // Render with specific initial state
    const { store } = await renderWithStore(<App />, {
      user: { name: "Toto" },
    });
    // Check called functions
    expect(login).toHaveBeenCalledTimes(1);
    expect(login.mock.calls[0][0]).toEqual("Toto"); // First argument of first call
    expect(loadGame).toHaveBeenCalledWith("FAKE_TOKEN", store.dispatch);
    expect(relogin).not.toHaveBeenCalled();
  });

  // Check auth process
  test("renders with stored token", async () => {
    // Render with data in localStorage
    localStorage.setItem("token", "FAKE_TOKEN");
    const { store } = await renderWithStore(<App />);
    // Check called functions
    expect(relogin).toHaveBeenCalledTimes(1);
    expect(relogin.mock.calls[0][0]).toEqual("FAKE_TOKEN"); // First argument of first call
    expect(loadGame).toHaveBeenCalledWith("FAKE_TOKEN", store.dispatch);
    expect(login).not.toHaveBeenCalled();
  });

  // Check auth process
  test("fill login form", async () => {
    // Render with specific initial state
    const { container } = await renderWithStore(<App />);
    // Fill form
    const input = screen.getByPlaceholderText("Your name");
    userEvent.type(input, "Toto");
    // Submit form
    const submit = screen.getByText("Log in");
    userEvent.click(submit);
    // Game should now be loading
    expect(screen.getByText("Loadingâ€¦")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
