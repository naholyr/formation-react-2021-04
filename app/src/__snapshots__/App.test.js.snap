// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App fill login form 1`] = `
<div>
  <h1>
    Motux
  </h1>
  <div>
    Loadingâ€¦
  </div>
</div>
`;

exports[`App snapshot 1`] = `
<div>
  <h1>
    Motux
  </h1>
  <form
    class="LoginForm"
  >
    <input
      minlength="2"
      name="username"
      placeholder="Your name"
      required=""
    />
    <button>
      Log in
    </button>
  </form>
</div>
`;
